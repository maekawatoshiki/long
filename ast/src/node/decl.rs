use super::{
    expr::Expr,
    stmt::BlockStmt,
    ty::{FuncType, Type},
};

/// A declaration kind.
/// <https://timsong-cpp.github.io/cppwp/n3337/dcl.dcl#1>
#[derive(Debug, Clone)]
pub enum Decl {
    FuncDef(FuncDef),
    SimpleDecl(SimpleDecl),
}

/// A function definition.
/// <https://timsong-cpp.github.io/cppwp/n3337/dcl.fct.def.general#nt:function-definition>
#[derive(Debug, Clone)]
pub struct FuncDef {
    pub name: DeclaratorId,
    pub ty: FuncType,
    pub body: BlockStmt,
}

/// A simple declaration.
/// <https://timsong-cpp.github.io/cppwp/n3337/dcl.dcl#nt:simple-declaration>
#[derive(Debug, Clone)]
pub struct SimpleDecl {
    pub names: Vec<DeclaratorId>,
    pub types: Vec<Type>,
    pub inits: Vec<Option<Expr>>,
}

/// A declarator id.
/// <https://timsong-cpp.github.io/cppwp/n3337/dcl.decl#nt:declarator-id>
#[derive(Debug, Clone)]
pub enum DeclaratorId {
    Ident(String),
}

impl From<FuncDef> for Decl {
    fn from(func_def: FuncDef) -> Self {
        Decl::FuncDef(func_def)
    }
}
