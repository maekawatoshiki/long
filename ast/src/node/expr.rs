use super::lit::Literal;
use super::Located;
use crate::token::kind::IntKind;

/// An expression kind.
#[derive(Debug, Clone)]
pub enum Expr {
    /// A literal expression.
    Literal(Literal),

    /// A unary expression.
    Unary(UnaryOp, Box<Located<Expr>>),

    /// A binary expression.
    Binary(BinOp, Box<Located<Expr>>, Box<Located<Expr>>),

    /// A ternary expression.
    Ternary(Box<Located<Expr>>, Box<Located<Expr>>, Box<Located<Expr>>), // cond, then, else.

    /// An Assign expression.
    Assign(AssignOp, Box<Located<Expr>>, Box<Located<Expr>>),
}

/// A unary operator kind.
#[derive(Debug, Clone, PartialEq)]
pub enum UnaryOp {
    Not,
}

/// A binary operator kind.
#[derive(Debug, Clone, PartialEq)]
pub enum BinOp {
    Comma,
    LogicalOr,
    LogicalAnd,
    And,
    Eq,
    Ne,
    Or,
    Gt,
    Ge,
    Lt,
    Le,
    Shl,
    Shr,
    Add,
    Sub,
    Mul,
    Div,
    Rem,
}
/// An assignment operator kind.
/// `=  *=  /=  %=   +=  -=  >>=  <<=  &=  ^=  |=`
/// <https://timsong-cpp.github.io/cppwp/n3337/expr.ass#1>
#[derive(Debug, Clone, PartialEq)]
pub enum AssignOp {
    Add,
    Sub,
    Mul,
    Div,
    Rem,
    Shl,
    Shr,
    And,
    Or,
    Xor,
    None,
}

impl Located<Expr> {
    // TODO: Add tests for this.
    /// Evaluates the constant expression.
    pub fn eval_constexpr(&self) -> Option<i64> {
        match self.inner_ref() {
            Expr::Unary(UnaryOp::Not, ref val) => Some((val.eval_constexpr() == Some(0)) as i64),
            Expr::Literal(Literal::Int(IntKind::Int(i))) => Some(*i as i64),
            Expr::Literal(_) => todo!(),
            Expr::Binary(BinOp::Comma, _, ref rhs) => rhs.eval_constexpr(),
            Expr::Binary(BinOp::LogicalOr, ref lhs, ref rhs) => {
                if lhs.eval_constexpr() != Some(0) {
                    Some(1)
                } else {
                    Some((rhs.eval_constexpr() != Some(0)) as i64)
                }
            }
            Expr::Binary(BinOp::LogicalAnd, ref lhs, ref rhs) => {
                if lhs.eval_constexpr() == Some(0) {
                    Some(0)
                } else {
                    Some((rhs.eval_constexpr() != Some(0)) as i64)
                }
            }
            Expr::Binary(ref op, ref lhs, ref rhs) => {
                let lhs = lhs.eval_constexpr()?;
                let rhs = rhs.eval_constexpr()?;
                match op {
                    BinOp::And => Some(lhs & rhs),
                    BinOp::Or => Some(lhs | rhs),
                    BinOp::Shl => Some(lhs << rhs),
                    BinOp::Shr => Some(lhs >> rhs),
                    BinOp::Eq => Some((lhs == rhs) as i64),
                    BinOp::Ne => Some((lhs != rhs) as i64),
                    BinOp::Gt => Some((lhs > rhs) as i64),
                    BinOp::Ge => Some((lhs >= rhs) as i64),
                    BinOp::Lt => Some((lhs < rhs) as i64),
                    BinOp::Le => Some((lhs <= rhs) as i64),
                    BinOp::Add => Some(lhs + rhs),
                    BinOp::Sub => Some(lhs - rhs),
                    BinOp::Mul => Some(lhs * rhs),
                    BinOp::Div => Some(lhs / rhs),
                    BinOp::Rem => Some(lhs % rhs),
                    BinOp::Comma => unreachable!(),
                    BinOp::LogicalAnd => unreachable!(),
                    BinOp::LogicalOr => unreachable!(),
                }
            }
            Expr::Ternary(ref cond, ref thn, ref els) => {
                if cond.eval_constexpr() != Some(0) {
                    thn.eval_constexpr()
                } else {
                    els.eval_constexpr()
                }
            }
            Expr::Assign(AssignOp::None, _, ref rhs) => rhs.eval_constexpr(),
            Expr::Assign(_, _, _) => todo!(),
        }
    }
}
