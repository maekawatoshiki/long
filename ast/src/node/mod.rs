use long_sourceloc::SourceLoc;

pub mod decl;
pub mod expr;
pub mod lit;
pub mod stmt;
pub mod ty;

/// Any kind of node with source location.
#[derive(Debug, Clone)]
pub struct Located<T> {
    /// Any kind of node.
    pub inner: T,

    /// The source location.
    pub loc: SourceLoc,
}

impl<T> Located<T> {
    /// Creates a new `Located`.
    pub fn new(inner: T, loc: SourceLoc) -> Self {
        Located { inner, loc }
    }

    /// Returns the inner node.
    pub fn inner(self) -> T {
        self.inner
    }

    /// Returns the reference to the inner node.
    pub fn inner_ref(&self) -> &T {
        &self.inner
    }

    /// Returns the source location.
    pub fn loc(&self) -> SourceLoc {
        self.loc
    }
}
