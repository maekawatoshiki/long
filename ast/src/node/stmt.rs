/// A statement kind.
#[derive(Debug, Clone)]
pub enum Stmt {
    Decl(Decl),
}

/// A declaration statement kind.
/// <https://timsong-cpp.github.io/cppwp/n3337/dcl.dcl#1>
#[derive(Debug, Clone)]
pub enum Decl {
    FuncDef(FuncDef),
}

/// A function definition.
/// <https://timsong-cpp.github.io/cppwp/n3337/dcl.fct.def.general#nt:function-definition>
#[derive(Debug, Clone)]
pub struct FuncDef {}
