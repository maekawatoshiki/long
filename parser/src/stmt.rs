use crate::{
    lexer::{traits::LexerLike, Error},
    Parser,
};
use anyhow::Result;
use long_ast::{
    node::{
        decl::Decl,
        expr::Expr,
        stmt::{BlockStmt, Stmt},
        Located,
    },
    token::kind::{KeywordKind, SymbolKind, TokenKind},
};

impl<'a, L: LexerLike> Parser<'a, L> {
    pub(crate) fn parse_function_body(&mut self) -> Result<BlockStmt> {
        self.parse_block()
    }

    /// Parses a statement.
    pub(crate) fn parse_stmt(&mut self) -> Result<Located<Stmt>> {
        let tok = self.lexer.next()?.ok_or_else(|| Error::UnexpectedEof)?;
        match tok.kind {
            TokenKind::Symbol(SymbolKind::OpeningBrace) => self
                .parse_block()
                .map(|x| Located::new(Stmt::Block(x), *tok.loc())),
            TokenKind::Keyword(KeywordKind::Return) => self
                .parse_return()
                .map(|x| Located::new(Stmt::Return(x), *tok.loc())),
            TokenKind::Keyword(kwd) if kwd.is_type() => match self.parse_decl(false)? {
                Located {
                    inner: Decl::SimpleDecl(decls),
                    loc,
                } => Ok(Located::new(Stmt::SimpleDecl(decls), loc)),
                _ => unreachable!(),
            },
            _ => {
                let loc = *tok.loc();
                self.lexer.unget(tok);
                let expr = self.parse_expr()?;
                self.expect(SymbolKind::Semicolon)?;
                Ok(Located::new(Stmt::Expr(expr), loc))
            }
        }
    }

    /// Parses a block statement. Assumes that '{' has already been skipped.
    fn parse_block(&mut self) -> Result<BlockStmt> {
        let mut stmts = vec![];
        loop {
            if self.lexer.skip(SymbolKind::ClosingBrace.into()) {
                break;
            }
            let stmt = self.parse_stmt()?;
            stmts.push(stmt);
        }
        Ok(BlockStmt(stmts))
    }

    /// Parses a return statement. Assumes that 'return' has already been skipped.
    fn parse_return(&mut self) -> Result<Option<Located<Expr>>> {
        if self.lexer.skip(SymbolKind::Semicolon.into()) {
            return Ok(None);
        }
        let expr = self.parse_expr()?;
        self.expect(SymbolKind::Semicolon)?;
        Ok(Some(expr))
    }
}

#[test]
fn parse_block() {
    use crate::lexer::Lexer;
    let node = Parser::new(&mut Lexer::new("{}")).parse_stmt();
    insta::assert_debug_snapshot!(node);
}

#[test]
fn parse_block2() {
    use crate::lexer::Lexer;
    let node = Parser::new(&mut Lexer::new("  { \n}      ")).parse_stmt();
    insta::assert_debug_snapshot!(node);
}

#[test]
fn parse_return() {
    use crate::lexer::Lexer;
    let node = Parser::new(&mut Lexer::new("return 0;")).parse_stmt();
    insta::assert_debug_snapshot!(node);
}

#[test]
fn parse_return2() {
    use crate::lexer::Lexer;
    let node = Parser::new(&mut Lexer::new("return;")).parse_stmt();
    insta::assert_debug_snapshot!(node);
}

#[test]
fn parse_return3() {
    use crate::lexer::Lexer;
    let node = Parser::new(&mut Lexer::new("return")).parse_stmt();
    insta::assert_debug_snapshot!(node);
}

#[test]
fn parse_vardecl() {
    use crate::lexer::Lexer;
    let node = Parser::new(&mut Lexer::new("int i;")).parse_stmt();
    insta::assert_debug_snapshot!(node);
}

#[test]
fn parse_vardecl2() {
    use crate::lexer::Lexer;
    let node = Parser::new(&mut Lexer::new("int i, j;")).parse_stmt();
    insta::assert_debug_snapshot!(node);
}
