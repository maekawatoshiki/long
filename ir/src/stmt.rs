use crate::expr::Expr;
use long_ast::node::Located;

/// A statement kind.
#[derive(Debug)]
pub enum Stmt<'a> {
    Expr(Located<&'a Expr<'a>>),
    Block(BlockStmt<'a>),
    Return(Option<Located<&'a Expr<'a>>>),
}

/// A block statement.
#[derive(Debug)]
pub struct BlockStmt<'a>(pub Vec<Located<&'a Stmt<'a>>>);
