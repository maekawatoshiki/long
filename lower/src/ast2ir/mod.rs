mod stmt;

use anyhow::Result;
use long_ast::node::{
    decl::{DeclaratorId, FuncDef},
    ty::{FuncType, Sign, Type as TypeNode},
};
use long_ir::{
    func::{Function, FunctionSignature},
    name::NameId,
    ty::Type as IrType,
    Module,
};
use stmt::lower_block;

/// A context used in the lowering process.
pub struct Context<'a> {
    module: &'a mut Module,
}

impl<'a> Context<'a> {
    /// Creates a new `Context`.
    pub fn new(module: &'a mut Module) -> Self {
        Self { module }
    }
}

/// Converts the given `FuncDef` into a `Function`.
pub fn lower_function(
    ctx: &mut Context,
    FuncDef {
        name,
        ty: FuncType { ret },
        body,
    }: FuncDef,
) -> Result<Function> {
    let name = resolve_declarator_id(ctx, name);
    let ret = resolve_type(ctx, ret);
    let sig = FunctionSignature {
        ret,
        params: vec![],
    };
    let body = lower_block(ctx, body)?;
    Ok(Function { name, sig, body })
}

fn resolve_declarator_id(ctx: &mut Context, declarator_id: DeclaratorId) -> NameId {
    // TODO
    match declarator_id {
        // TODO: Check if `name` is a fully-qualified name.
        DeclaratorId::Ident(name) => ctx.module.name_arena.find_or_create_global(&[name]),
    }
}

fn resolve_type(_ctx: &mut Context, ty_node: TypeNode) -> IrType {
    match ty_node {
        TypeNode::Void => IrType::void(),
        TypeNode::Int(Sign::Signed) => IrType::signed_int(),
        TypeNode::Int(Sign::Unsigned) => IrType::unsigned_int(),
        _ => todo!(),
    }
}

#[test]
fn parse_and_lower() {
    use long_ast::node::{decl::Decl, Located};
    use long_parser::lexer::Lexer;
    use long_parser::Parser;
    let Located { inner, .. } = Parser::new(&mut Lexer::new("int main() { return 0; }"))
        .parse_program()
        .unwrap()
        .into_iter()
        .next()
        .unwrap();
    let mut module = Module::new();
    let mut ctx = Context::new(&mut module);
    let func_ir = lower_function(
        &mut ctx,
        match inner {
            Decl::FuncDef(funcdef) => funcdef,
        },
    );
    insta::assert_debug_snapshot!(func_ir)
}
