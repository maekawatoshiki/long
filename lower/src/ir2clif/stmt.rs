use crate::ir2clif::{expr::lower_expr, FuncLowerCtx};
use anyhow::Result;
use cranelift::prelude::InstBuilder;
use cranelift_codegen::ir::types as clif_ty;
use long_ast::node::Located;
use long_ir::stmt::{BlockStmt, Stmt};

pub fn lower_stmt(ctx: &mut FuncLowerCtx, stmt: &Stmt) -> Result<()> {
    match stmt {
        Stmt::Block(_) => todo!(),
        Stmt::Expr(e) => {
            let _expr = lower_expr(ctx, &e.inner)?;
        }
        Stmt::Return(None) => {
            let ret = ctx.builder.ins().iconst(clif_ty::I32, 0);
            ctx.builder.ins().return_(&[ret]);
        }
        Stmt::Return(Some(Located {
            inner: expr,
            loc: _,
        })) => {
            let ret = lower_expr(ctx, expr)?;
            ctx.builder.ins().return_(&[ret]);
        }
    };
    Ok(())
}

pub fn lower_block_stmt<'a>(
    ctx: &'a mut FuncLowerCtx<'a, '_>,
    block: &'a BlockStmt<'a>,
) -> Result<()> {
    for stmt in block.0.iter() {
        lower_stmt(ctx, stmt.inner)?;
    }
    Ok(())
}
