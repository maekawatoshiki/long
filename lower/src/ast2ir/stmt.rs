use crate::ast2ir::{expr::lower_expr, Context};
use anyhow::Result;
use long_ast::node::{
    expr::Expr,
    stmt::{BlockStmt, Stmt},
    Located,
};
use long_ir::{
    block::{InstOrBlock, SimpleBlock},
    inst::{Inst, InstId},
};

pub(crate) fn lower_block(ctx: &mut Context, BlockStmt(stmts): BlockStmt) -> Result<SimpleBlock> {
    let mut seq = vec![];
    for Located {
        inner: stmt,
        loc: _, // TODO: We should not ignore this.
    } in stmts
    {
        match stmt {
            Stmt::Return(expr) => seq.push(InstOrBlock::Inst(lower_return(ctx, expr)?)),
            _ => todo!(),
        }
    }
    Ok(SimpleBlock(seq))
}

fn lower_return(ctx: &mut Context, expr: Option<Located<Expr>>) -> Result<InstId> {
    if expr.is_none() {
        return Ok(ctx.module.inst_arena.alloc(Inst::Return(None)));
    }

    let expr = expr.unwrap();
    let val = lower_expr(ctx, &expr.inner);
    Ok(ctx.module.inst_arena.alloc(Inst::Return(Some(val))))
}
