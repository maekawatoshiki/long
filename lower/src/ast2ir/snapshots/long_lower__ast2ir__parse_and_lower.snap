---
source: lower/src/ast2ir/mod.rs
expression: module

---
Module {
  name_arena: NameArena(Arena { arena_id: 3, items: [Global(["main"])], _phantom: PhantomData })
  val_arena: ValueArena(Arena { arena_id: 4, items: [Int(Int32(0))], _phantom: PhantomData })
  Function (::main) {
    sig: FunctionSignature { ret: Type(signed int), params: [] }
    body:
      Block:
        return Int(Int32(0))
  }
}
